File: src/update.ts
/**
 * Update checking module for DiffLearn
 * Checks GitHub releases for new versions and provides update instructions
 */

import * as fs from 'fs';
import * as path from 'path';

const GITHUB_REPO = 'lertsoft/DiffLearn';
const GITHUB_API_URL = `https://api.github.com/repos/${GITHUB_REPO}/releases/latest`;

export interface UpdateInfo {
    currentVersion: string;
    latestVersion: string;
    updateAvailable: boolean;
    releaseUrl: string;
    releaseNotes?: string;
    publishedAt?: string;
}

/**
 * Get the current version from package.json
 */
export function getCurrentVersion(): string {
    try {
        // Try to find package.json relative to this file or in common locations
        const possiblePaths = [
            path.join(__dirname, '../package.json'),
            path.join(__dirname, '../../package.json'),
            path.join(process.cwd(), 'package.json'),
        ];

        for (const pkgPath of possiblePaths) {
            if (fs.existsSync(pkgPath)) {
                const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf-8'));
                return pkg.version || '0.0.0';
            }
        }

        return '0.1.0'; // Fallback to known version
    } catch (_error: unknown) {
        return '0.1.0';
    }
}

/**
 * Fetch the latest release info from GitHub
 */
export async function getLatestRelease(): Promise<{
    version: string;
    url: string;
    notes?: string;
    publishedAt?: string;
} | null> {
    try {
        const response = await fetch(GITHUB_API_URL, {
            headers: {
                'Accept': 'application/vnd.github.v3+json',
                'User-Agent': 'DiffLearn-Update-Checker',
            },
        });

        if (!response.ok) {
            // No releases yet or rate limited
            return null;
        }

        const data = await response.json() as {
            tag_name?: string;
            html_url?: string;
            body?: string;
            published_at?: string;
        };

        // Tag name is usually "v0.1.0" format, strip the 'v' prefix
        const version = (data.tag_name || '').replace(/^v/, '');

        return {
            version,
            url: data.html_url || `https://github.com/${GITHUB_REPO}/releases`,
            notes: data.body,
            publishedAt: data.published_at,
        };
    } catch (_error: unknown) {
        // Network error or parsing error
        return null;
    }
}

/**
 * Compare two semver versions
 * Returns: 1 if v1 > v2, -1 if v1 < v2, 0 if equal
 */
function compareVersions(v1: string, v2: string): number {
    const parts1 = v1.split('.').map(Number);
    const parts2 = v2.split('.').map(Number);

    for (let i = 0; i < Math.max(parts1.length, parts2.length); i++) {
        const p1 = parts1[i] || 0;
        const p2 = parts2[i] || 0;

        if (p1 > p2) return 1;
        if (p1 < p2) return -1;
    }

    return 0;
}

/**
 * Check for updates and return update info
 */
export async function checkForUpdates(): Promise<UpdateInfo | null> {
    const currentVersion = getCurrentVersion();
    const latest = await getLatestRelease();

    if (!latest || !latest.version) {
        return null;
    }

    const updateAvailable = compareVersions(latest.version, currentVersion) > 0;

    return {
        currentVersion,
        latestVersion: latest.version,
        updateAvailable,
        releaseUrl: latest.url,
        releaseNotes: latest.notes,
        publishedAt: latest.publishedAt,
    };
}

/**
 * Determine if running from binary or source installation
 */
export function getInstallationType(): 'binary' | 'source' {
    // If we're running as a compiled binary, Bun.main will be the binary itself
    // For source installs, it will be the .tsx file
    const mainFile = process.argv[1] || '';

    if (mainFile.endsWith('.tsx') || mainFile.endsWith('.ts')) {
        return 'source';
    }

    return 'binary';
}

/**
 * Get the git root directory for source installations
 */
function getGitRoot(): string {
    try {
        // Look for package.json relative to this module
        const possibleRoots = [
            path.join(__dirname, '..'),
            path.join(__dirname, '../..'),
            process.cwd(),
        ];

        for (const root of possibleRoots) {
            if (fs.existsSync(path.join(root, 'package.json'))) {
                return root;
            }
        }
    } catch (_error: unknown) {
        // Fall through to default
    }
    return process.cwd();
}

/**
 * Get the appropriate update command based on installation type
 */
export function getUpdateCommand(): string {
    const installType = getInstallationType();

    if (installType === 'binary') {
        return `curl -fsSL https://raw.githubusercontent.com/${GITHUB_REPO}/master/install.sh | bash`;
    }

    // Source installation - use git pull from git root
    const gitRoot = getGitRoot();
    return `cd ${gitRoot} && git pull && bun install`;
}

/**
 * Format update info for display
 */
export function formatUpdateMessage(info: UpdateInfo): string {
    if (!info.updateAvailable) {
        return `✅ You're on the latest version (v${info.currentVersion})`;
    }

    const installType = getInstallationType();
    const updateCmd = getUpdateCommand();

    let message = `🆕 Update available: v${info.currentVersion} → v${info.latestVersion}\n\n`;

    if (installType === 'binary') {
        message += `To update, run:\n  ${updateCmd}\n`;
    } else {
        message += `To update (source install):\n  git pull && bun install\n`;
    }

    message += `\n📋 Release: ${info.releaseUrl}`;

    if (info.publishedAt) {
        const date = new Date(info.publishedAt).toLocaleDateString();
        message += `\n📅 Published: ${date}`;
    }

    return message;
}

File: go-source/internal/update/update.go
package update

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"runtime/debug"
	"strings"
)

const githubRepo = "lertsoft/DiffLearn"

func GetCurrentVersion() string {
	if info, ok := debug.ReadBuildInfo(); ok && info.Main.Version != "" && info.Main.Version != "(devel)" {
		return info.Main.Version
	}
	return "0.3.0"
}

type UpdateInfo struct {
	CurrentVersion string `json:"currentVersion"`
	LatestVersion  string `json:"latestVersion"`
	UpdateAvailable bool  `json:"updateAvailable"`
	ReleaseURL     string `json:"releaseUrl"`
	PublishedAt    string `json:"publishedAt,omitempty"`
}

func CheckForUpdates() (*UpdateInfo, error) {
	req, _ := http.NewRequest(http.MethodGet, "https://api.github.com/repos/"+githubRepo+"/releases/latest", nil)
	req.Header.Set("Accept", "application/vnd.github+json")
	req.Header.Set("User-Agent", "DiffLearn-Go")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		return nil, fmt.Errorf("github status: %d", resp.StatusCode)
	}
	var p struct {
		TagName     string `json:"tag_name"`
		HTMLURL     string `json:"html_url"`
		PublishedAt string `json:"published_at"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&p); err != nil {
		return nil, err
	}
	latest := strings.TrimPrefix(p.TagName, "v")
	current := GetCurrentVersion()
	return &UpdateInfo{CurrentVersion: current, LatestVersion: latest, UpdateAvailable: compareVersions(latest, current) > 0, ReleaseURL: p.HTMLURL, PublishedAt: p.PublishedAt}, nil
}

func compareVersions(v1, v2 string) int {
	var a1, b1, c1 int
	var a2, b2, c2 int
	fmt.Sscanf(v1, "%d.%d.%d", &a1, &b1, &c1)
	fmt.Sscanf(v2, "%d.%d.%d", &a2, &b2, &c2)
	if a1 != a2 {
		if a1 > a2 { return 1 }
		return -1
	}
	if b1 != b2 {
		if b1 > b2 { return 1 }
		return -1
	}
	if c1 != c2 {
		if c1 > c2 { return 1 }
		return -1
	}
	return 0
}

func GetUpdateCommand() string {
	exe := os.Args[0]
	if strings.HasSuffix(exe, ".go") {
		wd, _ := os.Getwd()
		return fmt.Sprintf("cd %s && git pull", filepath.Clean(wd))
	}
	return fmt.Sprintf("curl -fsSL https://raw.githubusercontent.com/%s/master/install.sh | bash", githubRepo)
}
